### PR TITLE
Optimise batchExecute with a batch size of 1000

### PR DESCRIPTION
**Pull request by kalaiyarasiganeshalingam**
## Purpose
$Subject

Fixes: https://github.com/ballerina-platform/ballerina-standard-library/issues/4129

## Checklist
- [x] Linked to an issue
- [ ] Updated the specification
- [ ] Updated the changelog
- [ ] Added tests
- [ ] Checked native-image compatibility

### CODE REVIEW

File: native/src/main/java/io/ballerina/stdlib/sql/nativeimpl/ExecuteProcessor.java [hunk 1]
```diff
@@ -183,10 +183,11 @@ private static Object nativeBatchExecuteExecutable(BObject client, BArray paramS
             }
             Connection connection = null;
             PreparedStatement statement = null;
-            ResultSet resultSet = null;
             String sqlQuery = null;
             List<Object[]> parameters = new ArrayList<>();
             List<BMap<BString, Object>> executionResults = new ArrayList<>();
+            boolean isGeneratedKeys = false;
```
Comment (niveathika) [inline native/src/main/java/io/ballerina/stdlib/sql/nativeimpl/ExecuteProcessor.java:189]:
```suggestion
            boolean processResultSet = false;
```

File: native/src/main/java/io/ballerina/stdlib/sql/nativeimpl/ExecuteProcessor.java [hunk 2]
```diff
@@ -208,28 +209,18 @@ private static Object nativeBatchExecuteExecutable(BObject client, BArray paramS
...
-                    resultSet = statement.getGeneratedKeys();
+                    isGeneratedKeys = true;
                 }
-                for (int count : counts) {
-                    Map<String, Object> resultField = new HashMap<>();
-                    resultField.put(Constants.AFFECTED_ROW_COUNT_FIELD, count);
-                    Object lastInsertedId = null;
-                    if (resultSet != null && resultSet.next()) {
-                        lastInsertedId = getGeneratedKeys(resultSet);
+                for (int i = 0; i < parameters.size(); i++) {
```
Comment (niveathika) [inline native/src/main/java/io/ballerina/stdlib/sql/nativeimpl/ExecuteProcessor.java:215]:
```suggestion
                for (int paramIndex = 0; paramIndex < parameters.size(); paramIndex++) {
```

File: native/src/main/java/io/ballerina/stdlib/sql/nativeimpl/ExecuteProcessor.java [hunk 3]
```diff
@@ -253,7 +244,7 @@ private static Object nativeBatchExecuteExecutable(BObject client, BArray paramS
                 return ErrorGenerator.getSQLError(th,
                         String.format("Error while executing batch command starting with: '%s'. ", sqlQuery));
             } finally {
-                closeResources(isWithinTrxBlock, resultSet, statement, connection);
+                closeResources(isWithinTrxBlock, null, statement, connection);
```
Comment (niveathika) [inline native/src/main/java/io/ballerina/stdlib/sql/nativeimpl/ExecuteProcessor.java:247]: Lets add a comment where the result set is cleaned for future reference.

File: native/src/main/java/io/ballerina/stdlib/sql/nativeimpl/ExecuteProcessor.java [hunk 4]
```diff
@@ -208,28 +209,18 @@ private static Object nativeBatchExecuteExecutable(BObject client, BArray paramS
...
+                    statement.addBatch();
+                    if ((i + 1) % batchSize == 0) {
+                        executeBatch(statement, executionResults, isGeneratedKeys);
+                        statement.clearBatch();
                     }
-                    resultField.put(Constants.LAST_INSERTED_ID_FIELD, lastInsertedId);
-                    executionResults.add(ValueCreator.createRecordValue(ModuleUtils.getModule(),
-                            Constants.EXECUTION_RESULT_RECORD, resultField));
                 }
+                executeBatch(statement, executionResults, isGeneratedKeys);
```
Comment (niveathika) [inline native/src/main/java/io/ballerina/stdlib/sql/nativeimpl/ExecuteProcessor.java:223]:
```suggestion
                // Execute leftover statements if count is not multiplier of baychSize
                executeSingleBatch(statement, executionResults, isGeneratedKeys);
```

**Review by daneshk — APPROVED** [review]
LGTM. @niveathika please check and approve

File: native/src/main/java/io/ballerina/stdlib/sql/nativeimpl/ExecuteProcessor.java (hunk 2 above)
Comment (niveathika) [reply native/src/main/java/io/ballerina/stdlib/sql/nativeimpl/ExecuteProcessor.java]: Still have't changed the variable name

**Review by niveathika — APPROVED** [review]
LGTM. Lets fix the changelog and specification

File: native/src/main/java/io/ballerina/stdlib/sql/nativeimpl/ExecuteProcessor.java (hunk 2 above)
Comment (kalaiyarasiganeshalingam) [reply native/src/main/java/io/ballerina/stdlib/sql/nativeimpl/ExecuteProcessor.java]: Updated in [784c70d](https://github.com/ballerina-platform/module-ballerina-sql/pull/629/commits/784c70d07880677e0082654dc2caa3fe107cfe7a)

File: docs/spec/spec.md [hunk 5]
```diff
@@ -459,7 +459,9 @@ sql:ExecutionResult result = check dbClient->execute(query);
 
 ## 4.4. Batch execute
 
-`batchExecute()` remote method executes the SQL query with multiple sets of parameters in a batch.
+`batchExecute()` remote method executes the SQL query with multiple sets of parameters in a batch. 
+To avoid the issues related to memory limitation, if the input data is more than 1000, 
+it will execute every 1000 records, concatenate those outputs and return it.
```
Comment (niveathika) [inline docs/spec/spec.md:464]:
```suggestion
To optimize memory utilisation, the inputs would be executed in batch sizes of 1000. 
```

**Review by niveathika — APPROVED** [review]

**Pull request merged**